src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1063 @@
+						}
+					}
+					else if(m_video_enable) // attribute mode
+					{
+						if(vram_data&0x80)
+						{
+							attribute = (vram_data&0x70)>>4;
+						}
+						else
+						{
+							if(attribute == 2)
+							{
+								// Low intensity
+								extra_intensity = 0;
+							}
+							else if(attribute == 3 && blink_strobe && !draw_cursor)
+							{
+								// Blink
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+							else if(attribute == 4)
+							{
+								// Inverse
+								data = ~data;
+								dot_nr8 = ~dot_nr8;
+							}
+							else if(attribute == 5 && line == 13)
+							{
+								// Underline
+								data = 0xff;
+								dot_nr8 = 0x01;
+							}
+							else if(attribute == 6)
+							{
+								// Invissible
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+						}
+					}
+
+					if(draw_cursor)
+					{
+						data = ~data;
+						dot_nr8 = ~dot_nr8;
+					}
+				}
+				if(full_inverse_video)
+				{
+					data = ~data;
+					dot_nr8 = ~dot_nr8;
+				}
+				for(int dot = 0; dot < 8; dot++)
+				{
+					bitmap.pix(chr_y_pos + line, chr_x_pos + dot) = palette[BIT(data, 7 - dot)<<extra_intensity];
+				}
+				bitmap.pix(chr_y_pos + line, chr_x_pos + 8) = palette[BIT(dot_nr8, 0)<<extra_intensity];
+			}
+		}
+	}
+	return 0;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// UART
+//
+
+void tandberg_tdv2100_disp_logic_device::update_rs232_lamps()
+{
+	m_write_waitl_cb(m_data_terminal_ready && !m_rx_handshake);
+	bool need_tx_for_on_line_led = ((m_dsw_u39->read()&0x0c0) == 0x080);
+	m_write_onlil_cb(m_data_terminal_ready && m_rx_handshake && (m_tx_handshake || !need_tx_for_on_line_led));
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_rxd_w(int state)
+{
+	if(!(m_data_terminal_ready && m_rx_handshake))
+	{
+		state = 1;
+	}
+	bool ext_echo = BIT(m_sw_rs232_settings->read(), 3);
+	if(!(ext_echo && m_data_terminal_ready))
+	{
+		state = (state && m_uart->so_r());
+	}
+	m_uart->write_si(state);
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_dcd_w(int state)
+{
+	m_write_carl_cb(state);
+	bool dcd_handshake = ((m_dsw_u39->read()&0x00c) == 0x008);
+	if(dcd_handshake)
+	{
+		m_rx_handshake = state;
+	}
+	update_rs232_lamps();
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_dsr_w(int state)
+{
+	bool auto_rx_handshake = BIT(m_sw_rs232_settings->read(), 4);
+	bool dcd_handshake = ((m_dsw_u39->read()&0x00c) == 0x008);
+	if(!dcd_handshake && !auto_rx_handshake)
+	{
+		m_rx_handshake = state;
+	}
+	update_rs232_lamps();
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_cts_w(int state)
+{
+	bool auto_tx_handshake = BIT(m_sw_rs232_settings->read(), 5);
+	if(!auto_tx_handshake)
+	{
+		m_tx_handshake = state;
+	}
+	update_rs232_lamps();
+}
+
+void tandberg_tdv2100_disp_logic_device::rs232_txd_w(int state)
+{
+	bool ext_echo = BIT(m_sw_rs232_settings->read(), 3);
+	if(!(ext_echo && m_data_terminal_ready))
+	{
+		m_uart->write_si(state);
+	}
```

Comment:
This doesn’t match with the logic in `rs232_rxd_w`.  In that function, you’re setting SI low if either the SO output is low and internal echo is enabled, or the RxD input is low.  This will set SI high if the SO output is low and internal echo is enabled even if the RxD input is low at that time.

Reply:
The RxD and SO+Echo are AND'ed together for SI, but this particular condition was not originally considered because this case is only valid if data-collision happens, and you are likely to get a framing error in both cases anyways.

But, I can add it.